.github/workflows/build.yml

```diff
@@ -0,0 +1,119 @@
+name: Build
+
+on:
+  push:
+    branches:
+      - master
+  workflow_dispatch:
+  pull_request:
+    branches:
+      - master
+  schedule:
+    - cron: '0 4 * * *' # Runs every day at 4am: https://docs.github.com/en/actions/reference/events-that-trigger-workflows#scheduled-events-schedule
+
+jobs:
+
+  build:
+    name: ${{ matrix.os }}, Node.js v${{ matrix.node }}, Python ${{ matrix.python }}
+
+    strategy:
+      matrix:
+        os: [windows-2019, ubuntu-18.04, macos-10.15]
+        node: ['12.x']
+        python: ['2.x']
+        include:
+          - os: ubuntu-18.04
+            node: '12.x'
+            python: '3.x'
+            tests: 'skip'
+
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 60
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # To fetch all history for all branches and tags. (Will be required for caching with lerna: https://github.com/markuplint/markuplint/pull/111)
+
+      - name: Use Node.js ${{ matrix.node }}
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ matrix.node }}
+          registry-url: 'https://registry.npmjs.org'
+
+      - name: Use Python ${{ matrix.python }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python }}
+
+      - name: Build
+        shell: bash
+        run: |
+          yarn --skip-integrity-check --network-timeout 100000
+          npx electron-replace-ffmpeg
+          npx electron-codecs-test
+          ./scripts/check_git_status.sh
+        env:
+          NODE_OPTIONS: --max_old_space_size=4096
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # https://github.com/microsoft/vscode-ripgrep/issues/9
+
+      - name: Test (headless)
+        if: matrix.tests != 'skip'
+        shell: bash
+        run: |
+          yarn test:theia
+
+      - name: Test (browser)
+        if: matrix.tests != 'skip' && runner.os == 'Linux'
+        uses: GabrielBB/xvfb-action@v1
+        with:
+          run: yarn test:browser
+
+      - name: Test (electron)
+        if: matrix.tests != 'skip' && runner.os == 'Linux'
+        uses: GabrielBB/xvfb-action@v1
+        with:
+          run: yarn test:electron
+
+  publish:
+    needs: build
+    if: github.ref == 'refs/heads/master' && github.event_name != 'schedule' # We still publish the manually dispatched workflows: 'workflow_dispatch'.
+    runs-on: ubuntu-18.04
+
+    # The current approach is silly. We should be smarter and use `actions/upload-artifact` and `actions/download-artifact` instead of rebuilding
+    # everything from scratch again. (git checkout, Node.js install, yarn, etc.) It was not possible to share artifacts on Travis CI without an
+    # external storage (such as S3), so we did rebuild everything before the npm publish. We should overcome this limitation with GH Actions.
```

Comment:
I overcame this limitation on my PR by doing only one job that conditionally publishes as a last step. No artifact uploading weirdness required.

Reply:
> I overcame this limitation on my PR by doing only one job that conditionally publishes as a last step

I've seen that. I think that would end up in a flawed pipeline as publishing could happen on the Linux node right after the (green) tests step, but the Windows build job could have failed. This PR mimics the current Travis CI behavior: publishing happens only when all builds are green.

To achieve this👆, I usually use multiple jobs with `needs` to create a dependency between the jobs. According to the [`needs context`](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#needs-context) doc, this will ensure that the previous job has a `result` (`'success'`). So the publish happen only **iff** all builds where green. Do you have a better suggestion?

Reply:
Right. Thinking about it, to publish all we need is to transpile our TypeScript sources into JavaScript. We could get away just doing `yarn build` then? Except that we need to do `yarn` first and doing this does `install, build, lint, build example apps and download plugins` which is what annoys me the most with our `prepare` script right now. Do you think it'd be better to also just run the build step for the publish?